### ISSUE TITLE
macOS Mojave: Atom requests access to Contacts and Calendars

### ISSUE THREAD

**Issue by freshyill** [atom/atom#17687]
I [raised this issue](https://discuss.atom.io/t/atom-would-like-to-access-my-calendar-and-contacts/56924) on the forum and it was suggested that I file an issue. It's not a standard bug, so I can't really go through the steps to reproduce. It occurred on 1.28.1.

On a clean install of OSX Mojave Public Beta (immediately installed on a new machine), Atom requests access to contacts and calendars on first launch. It has never requested on a different machine running High Sierra. Obviously, since this is a new installation, there are no plugins or themes installed.

I denied both requests but I don’t see any reason why this is necessary. 

![6332729e4c23db4df7cf77c3e5a34e508b8eb33a_1_690x347](https://user-images.githubusercontent.com/879292/42726343-e552dcbe-8760-11e8-860a-1e034217b87d.png)

**Comment by rsese** [comment]
Thanks for the report - I don't have the beta install to test but it does seem like there's mention of this with Mojave from a quick search:

* https://www.dropboxforum.com/t5/Installation-and-desktop-app/dropbox-is-asking-for-calendar-permissions-now/td-p/281776
* https://github.com/facebook/osquery/issues/4492

This one was interesting too:

* https://www.felix-schwarz.org/blog/2018/06/apple-event-sandboxing-in-macos-mojave

> Apps that traverse the user's home folder could trigger multiple approval prompts, not just for Photos, Contacts, Calendars and so on.

**Comment by freshyill** [comment]
For what it's worth, I was able to repeat the behavior on a brand new Mojave beta on a brand new machine. 

<img width="532" alt="screen shot 2018-07-25 at 9 27 42 pm" src="https://user-images.githubusercontent.com/879292/43236514-7f51e2ec-9053-11e8-866e-d0029791cc88.png">
<img width="532" alt="screen shot 2018-07-25 at 9 27 31 pm" src="https://user-images.githubusercontent.com/879292/43236515-7f615f42-9053-11e8-96d5-6754ebbfde6b.png">

**Comment by ssbarnea** [comment]
I am not sure if this is related but in my case I had my home user folder as a git repository, which. makes me think that even my shell prompt could trigger this and any other tool using git, like atom, visual studio code,... 

I wonder if there is a workaround for this as it makes the OS almost impossible to use due to the stacking popups.

**Comment by donaldrelliott** [comment]
I'll add in that on the Mojave Beta Atom also requests to have access to my photos. For me those popups happen quite a bit so I've had to resort to using Microsoft Code instead for the time being.

**Comment by ssbarnea** [comment]
This is not really Atom fault but I was able to narrow it down. If you open you user home directory atom (or git it you added it as a git repo) will try to navigate into Applications Support folder amd trigger this “security” prompts. Any app that will search in files would trigger this, even a “find”.

**Comment by daviwil** [comment]
Thanks @ssbarnea! I found the issue you filed on VS Code for the same problem:

https://github.com/Microsoft/vscode/issues/55236

It does appear that this is due to the new privacy features that are being introduced in Mojave, Apple raises this prompt for *any* app that tries to read files in certain user folders.  For now there doesn't seem to be an understood resolution to this issue so we'll keep monitoring the situation in case a solution arises.

**Comment by queerviolet** [comment]
Is anyone still experiencing this with the latest Mojave + Atom? I'm on macOS 18A384a and Atom 1.30.0 and can't repro.

**Comment by queerviolet** [comment]
I've been trying without luck to repro on beta 10. Is anyone on the most recent beta still experiencing this?

**Comment by jsphkm** [comment]
I was able to reproduce these alerts (both calendar and contacts) with Atom in the public release of Mojave

**Comment by queerviolet** [comment]
How'd you do it? I'm on Mojave and don't see them.

**Comment by jsphkm** [comment]
I think the prompts came up because I've installed Mojave for the first time and haven't provided my response before.  So I don't think it is fixed, I think the app(or os) has remembered the previous response for those that are not able to reproduce the prompts anymore.

**Comment by jsphkm** [comment]
Those two prompts only came up once each when starting the program for the first time after the installation of Mojave.  They did not appear again afterwards in my experience.  Also, I did `code .` in my project folder.

**Comment by 50Wliu** [comment]
> Also, I did `code .` in my project folder.

`code` is for VSCode - are you saying that it also happened for VSCode as well as Atom?

**Comment by jsphkm** [comment]
oops my mistake, I was so used to visual code that i wrote in `code .`  What I meant was that I manually opened the file from the finder.  Also, I can't remember if it came up on Visual Code.

**Comment by jsphkm** [comment]
Oh, on a second thought, I think this has to do with the Atom packages.  I don't have many plugins in Visual Code, but I do have several in Atom.  I think one of those is triggering it, my bets are on Hydrogen.  I'm coming to this conclusion after finding that Hyper wants to access Google Chrome for React apps.
![screen shot 2018-09-25 at 12 02 20 pm](https://user-images.githubusercontent.com/28643797/46036666-edbf7e00-c0ba-11e8-81b8-7bef1607e180.png)

**Comment by NathanTheGr8** [comment]
I encounter this with atom 1.30.0 on mac OS 10.14 (18A391) open ~/.bash_profile

**Comment by noname2534** [comment]
ในวันที่ อา. 30 ก.ย. 2018 15:20 น. Dave Carlson <notifications@github.com>
เขียนว่า:

**Comment by jsphkm** [comment]
Ok, I was able to reproduce the prompts for VC by doing:
```
cd ~
code .
```
That brought up prompts for contacts, calendars, and photos.
However, I should note that I was initially working with folders & files inside `~/Code/projects/`
And with that directory path I first triggered those prompts with Atom and not Visual Code, hence why I posted.

**Comment by lee-dohm** [comment]
Reproduced using Atom Nightly v1.33.0-nightly2 on macOS 10.14.

**Comment by queerviolet** [comment]
Does it happen in safe mode? I can't repro it, so I'd suspect it's an extension.

**Comment by lee-dohm** [comment]
The first time I launched Atom Nightly v1.33.0-nightly2, it requested access to my photos. I declined it, so it showed up in Privacy settings unchecked. Between then and trying to reproduce this again, I updated to Atom Nightly v1.33.0-nightly5. It would not reproduce after that, which is expected behavior because macOS was remembering the choice from before.

I reset my Photos security choices using `tccutil reset Photos`, which deleted all of the security settings in Privacy for the Photos application. I have then attempted to reproduce this using both Atom Nightly v1.33.0-nightly5 and Atom v1.31.1 in both Safe Mode and normal mode and it does not reproduce.

**Comment by jasonrudolph** [comment]
I looked into this issue today. Here's what I'm seeing:

**tl;dr** You'll see these prompts the first time you perform an action that instructs Atom to traverse directories that have Mojave's new OS-level protections. Once you make your choice, Mojave won't bother you again. Because Mojave applies these same protections to all apps (even Apple's own `Terminal.app`), I think education is the most appropriate approach for addressing this issue. I recommend that we add a support article that explains why these prompts appear and that they'll only appear once.

### How to reproduce

I am able to reliably reproduce this issue with the following steps:

1. Install Mojave (macOS 10.14)
2. Install Atom. (I used 1.31.1, but I suspect any version will suffice.)
3. Launch Atom
4. Inside Atom, run the "Window: Install Shell Commands" command
5. From a terminal, run `$ atom ~/.profile`
6. Hit <kbd>Command</kbd>+<kbd>t</kbd> to open the fuzzy-finder

After a few moments, macOS shows a prompt saying that '"Atom" would like access to your calendar':

<img width="1030" alt="screen shot 2018-10-03 at 11 38 20 am" src="https://user-images.githubusercontent.com/2988/46432185-9bfe9f80-c71b-11e8-83f7-758c3212d16c.png">

### Why is this happening?

When you open the fuzzy-finder, Atom indexes the currently-open directory so that it can show you the available files. Because you're editing a file at the root of your home directory (`~/.profile`), the current directory is your home directory (`~`). Your home directory also contains your files that have new OS-level protections in Mojave:

- Calendar files (`~./Library/Calendars`)
- Contacts files (`~/Library/Application\ Support/AddressBook`
- Photos files (`~/Pictures/Photos\ Library.photoslibrary`)

There may be other categories of files that receive these new OS-level protections as well (Mail?, Reminders?, Time Machine backups?)

Before letting Atom read these files, Mojave is understandably asking whether you want Atom to be able to access this personal data.

### Reproducible in Apple's macOS apps also

Notably, macOS enforces these same protections even for some Apple-owned apps that Apple ships with macOS Mojave.

For example, if you open `Terminal.app` and try to list the files in `~/Library/Calendars`, macOS shows a prompt saying that '"Terminal" would like access to your calendar':

<img width="602" alt="screen shot 2018-10-03 at 14 04 40 pm" src="https://user-images.githubusercontent.com/2988/46432184-9bfe9f80-c71b-11e8-8916-e844b7f4110e.png">


### You're only prompted once

Fortunately, macOS will only prompt you once for each type of personal data. In other words, you might see a prompt asking you whether Atom can access your calendar, and you might see a prompt asking you whether Atom can access your contacts, but once you make those decisions, you won't see those prompts again.

Even when you upgrade to a new version of Atom, macOS remembers your previous choices and won't prompt you again.

At any time, you can change your choices via System Preferences. Go to `System Preferences -> Security and Privacy`, click the `Privacy` tab, and then click on `Calendars` to manage which apps can access your `Calendars`. The same goes for `Contacts`, `Photos`, etc.:

<img width="780" alt="screen shot 2018-10-03 at 2 53 58 pm" src="https://user-images.githubusercontent.com/2988/46432459-51315780-c71c-11e8-96b5-83edea8b4be2.png">

### Where do we go from here?

Since this new macOS behavior affects any third-party app that tries to list files in certain directories (and since it even affects some first-party apps like `Terminal.app`), I recommend that we publish a support article that explains _why_ these prompts appear, that these prompts affect all apps, that these prompts will only appear once.

### Possible alternatives

A few alternative approaches come to mind. I don't think we should invest time in these efforts, since doing so would take focus away from our other priorities, but I'll capture these alternatives here for posterity:

- We could teach fuzzy-finder (and any other part of Atom that traverses directories) to ignore these protected directories. That would prevent the prompts from appearing. However, if someone purposefully wanted to use the fuzzy-finder on those directories, they wouldn't be able to do so.
- We could teach fuzzy-finder (and any other part of Atom that traverses directories) to prompt you *before* it attempts to access these protected directories. By doing that, Atom could control the messaging, and it would reduce the surprise that users experience when macOS unexpectedly alerts the user that '"Atom" would like access to your calendar'.
- We could add a configuration option that instructs Atom to always ignore these protected directories. If this configuration option were enabled by default, Atom wouldn't attempt to read these protected directories, and therefore users wouldn't see the surprising macOS prompts.
- When opening the `~/Library` directory, the home directory (`~`), or any of its parent directories, Atom could alert you that you may see a prompt from macOS if you attempt to access any of the newly-protected files while using Atom.

Of course, each one of these alternatives brings tradeoffs that would need to be considered. At a glance, I think those tradeoffs would result in a poorer user experience overall.

---

I hope the information above is helpful. If any of these findings differ from the behavior you're seeing, please let me know. :bow:

**Comment by lee-dohm** [comment]
Additional alternatives:

-----

We could add those directories to [the default `core.ignoredNames` value](https://github.com/atom/atom/blob/04d95a76f8b98d5e0e0e0afc9c42de9603ca250e/src/config-schema.js#L10).

**Pros:** This will respect people's already-configured defaults for things like whether fuzzy-finder includes or excludes ignored names. Fast, easy change.
**Cons:** People who have customized their `core.ignoredNames` value will not get this change

-----

We could add those directories to [the default `fuzzy-finder.ignoredNames` value](https://github.com/atom/fuzzy-finder/blob/f096a3e1faea35573d1619e410496031e9cccce4/package.json#L60).

**Pros:** This will respect people's already-configured defaults for things like whether fuzzy-finder includes or excludes ignored names. Fast, easy change.
**Cons:** People who have customized their `fuzzy-finder.ignoredNames` value will not get this change

**Comment by jasonrudolph** [comment]
> Because Mojave applies these same protections to all apps (even Apple's own `Terminal.app`), I think education is the most appropriate approach for addressing this issue. I recommend that we add a support article that explains why these prompts appear and that they'll only appear once. -- https://github.com/atom/atom/issues/17687#issuecomment-426760941

We've published a support article explaining why these prompts appear and offering options for handling them: https://discuss.atom.io/t/why-does-macos-say-that-atom-wants-to-access-my-calendar-contacts-photos-etc/59291

With that article in place, I'm going to close out this issue.

---

@freshyill: Thank you for bringing this to our attention. :bow:

**Comment by 1pocketaces1** [comment]
Hi my Atom editor no longer auto-completes or has proper coloring after switching to Mojave. Is there any known cause or fix to this?
